dask/array/core.py

```diff
@@ -2233,6 +2233,259 @@ def unpack_singleton(x):
     return x
 
 
+class _Recurser(object):
```

Comment:
Code-style-wise things like this concern me because it increases the depth that a new reader needs to understand before being able to access a particular piece of code.  However I see that this has just been lifted from the NumPy codebase, so presumably we're not adding much on top of what is already present.

Are we able to import this rather than re-implement it?  Presumably this would only work with NumPy versions where this already existed, but that might be ok (we shouldn't necessarily back-port numpy functionality in dask.array (or should we?)).

Reply:
Agreed that it is not great. Unfortunately this is not part of NumPy's public API so we don't really have lots of options here. Importing it doesn't avoid maintenance burden it just shifts it to dealing with unpredictable breaks that newer versions of NumPy may cause if this class changes. Not to mention this is only present in NumPy 1.13 so we would then require the user to have the newest NumPy version to get `block`. Seems easier to me to just copy it (not to say that I like it one bit).

Reply:
Should add there might be some more Dask-like way to accomplish this as well, but I wasn't able to think of it at the time.

Reply:
If we're keeping this thing then it would be nice to hide it in a utils file

Reply:
Right. So was proposing we put this in `dask.array.numpy_compat`. Does that seem ok?

Reply:
Seems better than alternatives.